### DISCUSSION TITLE
Bit confused about void TFT_eSPI::drawGlyph(uint16_t code)

### DISCUSSION THREAD

**Discussion by Frank-Bemelman** [Bodmer/TFT_eSPI#1927] [General]
Trying to understand why my ili9341 240x320 display shows quite a bit of flickering when updating it very frequently (10 times a second) with text in smooth font, so I dug into this function just for the heck of it.

First part is to fill the background above the glyph. Ok.
Then the drawing of the glyph, walking through each y and x. Makes sense.
And finally fill the background below the glyph. Ok.

I see two nested for loops to draw the actual glyph:
    for (int32_t y = 0; y < gHeight[gNum]; y++)
and
    for (int32_t x = 0; x < gWidth[gNum]; x++)

There is a lot going  on here, that I don't understand. Pixel checks, bl and fl, drawFastHLine calls.

If I remove all code, and only keep these two lines:
            if (getColor) bg = getColor(x + cx, y + cy);
            drawPixel(x + cx, y + cy, alphaBlend(pixel, fg, bg));

...it works just as good (in my case anyway). Even when I leave out the getColor line, I see no diffence - but I could imagine that comes into place when printing text over an image background.

So I am a bit puzzled about all this code and it's purpose.

In the meantime, I am starting to believe the flickering I notice is an artefact of the display itself. It is also much more noticabe when I look at an angle, and much less when I look at it sitting perfectly in front. 

That's all folks ;-)

**Comment by Bodmer** [comment]
The flickering is likely due to the display being updated too frequently with changing values, or maybe the screen is being cleared between updates? Post a simple complete example demonstrating the problem.

The extra code queues up pixels of the same colour for more efficient (faster) rendering.

**Reply by Frank-Bemelman** [reply]
First I thought that too, even dived into your code to see if something was the culprit there - which was not the case. I do many updates with my horizontal scrolling of text, about 10 times a second. I am not using a push bitmap buffer, but just print the text over and over again. This allows me to use dynamic text that changes all the time while it goes. 

This flickering was with an 320x240 ILI9341 TFT display. With another display based on ST7789, IPS type, I have no visible flickering and everything goes smooth.  I think it is just something with the quality of the panel that my ILI9341 display uses. Too bad that there are no too many real amoled displays on the market , those probably perform even better dan IPS displays.

I could post an example but it is quite large. However, I did made a repository for my jukebox project in which it is used. Works with both a ILI9431 and ST7789. I also used your dial example to make a volume control, and since the ST7789 has no read-back, I simply paint the part under the needle using the jpg background image. It is slower, but not bad at all. Thanks to the speed of everything.

Added a small video. On the video is not so noticeable, but you can see it. I blame it on the particular display. The moire is caused by my iphone, it is not on the display iself. Here is the link:

https://www.youtube.com/watch?v=PuYiOFJjZX4
